src/core/net/srp_client.cpp

```diff
@@ -953,34 +946,180 @@ Error Client::ReadOrGenerateKey(Crypto::Ecdsa::P256::KeyPair &aKeyPair)
-        // renewal time of a registered service, we refresh the service
-        // early and include it in this update. This helps put more
-        // services on the same lease refresh schedule.
+        // When host is being removed, there is no need to include
+        // services in the message (server is expected to remove any
+        // previously registered services by this client). However, we
+        // still mark all services as if they are appended in the message
+        // so to ensure to update their state after sending the message.
+
+        for (Service &service : mServices)
+        {
+            service.MarkAsAppendedInMessage();
+        }
+
+        mLease    = 0;
+        mKeyLease = mShouldRemoveKeyLease ? 0 : mDefaultKeyLease;
+        ExitNow();
+    }
+
+    mLease    = kUnspecifiedInterval;
+    mKeyLease = kUnspecifiedInterval;
+
+    // We first go through all services which are being updated (in any
+    // of `...ing` states) and determine the lease and key lease intervals
+    // associated with them. By the end of the loop either of `mLease` or
+    // `mKeyLease` may be set or may still remain `kUnspecifiedInterval`.
+
+    for (Service &service : mServices)
+    {
+        uint32_t lease    = DetermineLeaseInterval(service.GetLease(), mDefaultLease);
+        uint32_t keyLease = Max(DetermineLeaseInterval(service.GetKeyLease(), mDefaultKeyLease), lease);
+
+        service.ClearAppendedInMessageFlag();
+
+        switch (service.GetState())
+        {
+        case kAdding:
+        case kRefreshing:
+            OT_ASSERT((mLease == kUnspecifiedInterval) || (mLease == lease));
+            mLease = lease;
+
+            OT_FALL_THROUGH;
+
+        case kRemoving:
+            OT_ASSERT((mKeyLease == kUnspecifiedInterval) || (mKeyLease == keyLease));
+            mKeyLease = keyLease;
+            break;
+
+        case kToAdd:
+        case kToRefresh:
+        case kToRemove:
+        case kRegistered:
+        case kRemoved:
+            break;
+        }
+    }
+
+    // We go through all services again and append the services that
+    // match the selected `mLease` and `mKeyLease`. If the lease intervals
+    // are not yet set, the first appended service will determine them.
+
+    for (Service &service : mServices)
+    {
+        // Skip over services that are already registered in this loop.
+        // They may be added from the loop below once the lease intervals
+        // are determined.
+
+        if ((service.GetState() != kRegistered) && CanAppendService(service))
+        {
+            SuccessOrExit(error = AppendServiceInstruction(service, aMessage, aInfo));
+
+            if (mSingleServiceMode)
+            {
+                // In "single service mode", we allow only one service
+                // to be appended in the message.
+                break;
+            }
+        }
+    }
+
+    if (!mSingleServiceMode)
```

Comment:
If the for-loop above does not append any service instructions, do we still want to avoid appending service instructions in single service mode here?

Reply:
Yes. This should be fine. Under this `if()` we are checking if we can do "opportunistic early refresh" of some already registered services. This won't really matter (or happen) when we are using `mSingleServiceMdoe`

In order to use `mSingleServiceMode`, we need to go through this a first round and prepare a message with all services and then determine that message size is larger than MTU and  then go through this one more round and only append one service. Any service that would be eligible for "opportunistic early refresh" would be determined in the first round (and its state would be changed to `kToRefresh` in the first round).

Reply:
In single service mode, do we really want to do any "opportunistic early refresh"? In other words, should we be setting state to `kToRefresh` if not everything fits in a single message?

Reply:
That's correct. Under `mSingleServiceMode` we won't. We have the `if (!mSingleServcieMode)` to skip.

But we may have decided to do "early refresh" of some services earlier (before knowing that things won't fit in the MTU message). This is similar to how it works with existing `main` code. 
- We can consider adding more checks to stop early if message is already above the MTU threshold. 
- It may be a good  enhancement. I would say maybe better to consider it in a separate PR (adding it if it does not add too much code size/complexity)

Reply:
Yes, I was thinking that on the first pass, we will set all "early refresh" services to `kToRefresh`, but it will stay that way if we later determine to enter single service mode. We do early refresh because we want to combine a single message. But it's not obvious we want to do that if each service needs its own message. If want to address it, I agree it can be done in a separate PR.

Reply:
Some ideas/thought for future changes possibly replacing/improving the "single service mode" (after this PR):
- We now try to add as many services and if msg gets larger than MTU, we enable "single service mode" and include only one service and then try again hoping the rest fit again (if not repeat the same).
- With this PR, we added mechanism to track which set of services are included in the prepared message.
- With this, we can change the model: 
   - Try to add as many services, if msg gets larger than MTU, we start from current set again and exclude services one by one until msg gets below MTU. Basically we try to fit as many services as we can in the first msg instead of including only one. 
   - If/when we go for this model, I think we may still keep and do the "early refresh".

Reply:
SGTM!